python/subunit/v2.py

```diff
@@ -415,6 +415,13 @@ def _parse(self, packet, result):
             length, consumed = self._parse_varint(
                 packet[-1], 2, max_3_bytes=True)
             remainder = self.source.read(length - 6)
+            for _ in range(5):
+                # workaround for bug #1813147 - if we haven't parsed enough
+                # input, try again in case we're racing
+                if len(remainder) == length - 6:
+                    break
+                remainder += self.source.read(length - len(remainder) - 6)
```

Comment:
I think this is incomplete: you'll also need to handle the case of reading an incomplete header on line 411.

*however* I don't think this is a good idea at all. The underlying problem here is that your captured file emulation is buggy. Fixing that should be pretty easy and then you won't race at all.

Reply:
> I think this is incomplete: you'll also need to handle the case of reading an incomplete header on line 411.

Good point

> *however* I don't think this is a good idea at all. The underlying problem here is that your captured file emulation is buggy. Fixing that should be pretty easy and then you won't race at all.

Yeah, to be clear, this is very much a workaround that exists because I can't identify the root cause of the issue among the trio of testtools, subunit and stestr. I'd prefer to figure that out but if I can't, this does achieve its goal of masking said bug effectively enough for me (it's killing us slowly in nova land)

Reply:
lets try to narrow that down. First - what is the type of the file object that is being supplied? And where is it being constructed?

We could alternatively interpret this as subunit being enhanced to work with non-blocking streams; for that though we'd also want to add in some sort of readiness callback (suitable for use with a reactor/proactor/poll etc), accumulate what was actually read for later evaluation (as you do here, but also for the header) and generally be friendly to such files

Reply:
> lets try to narrow that down. First - what is the type of the file object that is being supplied? And where is it being constructed?

It looks like `stestr run` is using `io.BytesIO` [1](https://github.com/mtreinish/stestr/blob/ff9b6887df2e0ce7688601ffcf4537aabe05dbdb/stestr/subunit_runner/run.py#L79) [2](https://github.com/mtreinish/stestr/blob/ff9b6887df2e0ce7688601ffcf4537aabe05dbdb/stestr/subunit_runner/run.py#L39) [3](https://github.com/mtreinish/stestr/blob/ff9b6887df2e0ce7688601ffcf4537aabe05dbdb/stestr/commands/run.py#L465). Looking at [the docs](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read), it seems `read` will "Read and return **up to** size bytes" so perhaps reading again is correct.

> We could alternatively interpret this as subunit being enhanced to work with non-blocking streams; for that though we'd also want to add in some sort of readiness callback (suitable for use with a reactor/proactor/poll etc), accumulate what was actually read for later evaluation (as you do here, but also for the header) and generally be friendly to such files

Sounds good, but my knowledge of Python's IO model isn't great so while I'm happy to do whatever's needed, I'll probably need a little help :smile:

Reply:
I've reworked this to address the header also. Not sure how best to add the readiness callback mechanism or whether it's necessary for this issue at hand.

Reply:
Thanks for digging up the references.